File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/card/infrastructure/scheduling/CardStatementScheduler.java
package com.example.accountservice.card.infrastructure.scheduling;

import com.example.accountservice.taskqueue.TaskOutboxWriter;
import java.time.YearMonth;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * Once a month, enqueues a task to notify all active cards of last month's usage (monthly card
 * usage statement dispatch, scheduling.md Feature 2). The Scheduler does not execute business logic
 * directly — it only enqueues a task to the Task Queue. The actual statistics calculation and
 * sending are handled by {@code card/interfaces/task/SendCardStatementTaskController} → {@code
 * SendCardStatementService}.
 */
@Component
@RequiredArgsConstructor
public class CardStatementScheduler {

    private static final Logger log = LoggerFactory.getLogger(CardStatementScheduler.class);
    private static final String TASK_TYPE = "card.send-statement";
    private static final String GROUP_ID = "card.statement";

    private final TaskOutboxWriter taskOutboxWriter;

    // With a month-based deduplicationId, even if multiple instances tick simultaneously on the
    // 1st of the same month, only one entry ends up in the Task Queue (scheduling.md "Cron
    // multi-instance safety"). Exceptions are only logged, not rethrown — the next tick (4 AM on
    // the 1st of the following month) will retry.
    @Scheduled(cron = "0 0 4 1 * *") // 4 AM on the 1st of every month
    public void enqueueMonthlyStatement() {
        try {
            YearMonth month = YearMonth.now();
            String dedupId = TASK_TYPE + "-" + month;
            taskOutboxWriter.enqueue(TASK_TYPE, new Payload(month), GROUP_ID, dedupId);
        } catch (Exception e) {
            log.error("Failed to enqueue card statement task", e);
        }
    }

    // The local payload view owned by this Scheduler — contracts only via the JSON schema (field
    // names) with the separate payload record owned by
    // card/interfaces/task/SendCardStatementTaskController (same reasoning as
    // account/infrastructure/scheduling/InterestPaymentScheduler).
    private record Payload(YearMonth month) {}
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/card/infrastructure/scheduling/CardStatementScheduler.kt
package com.example.accountservice.card.infrastructure.scheduling

import com.example.accountservice.taskqueue.TaskQueue
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.YearMonth

/**
 * The Scheduler for monthly card-usage-statement delivery — Infrastructure layer (the same
 * role/structure as
 * [com.example.accountservice.account.infrastructure.scheduling.InterestPaymentScheduler]). It only
 * enqueues the Task onto the [TaskQueue]; the actual aggregation/delivery is handled by
 * [com.example.accountservice.card.interfaces.task.SendCardStatementTaskController] →
 * `SendMonthlyCardStatementsService`, which receive the `card.send-statement` Task.
 *
 * [yearMonth] (the current month at enqueue time) is used only as the duplicate-send-prevention key —
 * the actual aggregation period (the last 30 days) is calculated separately by
 * `SendMonthlyCardStatementsService` based on when the Task is processed (see that class's KDoc).
 */
@Component
class CardStatementScheduler(
    private val taskQueue: TaskQueue,
    private val objectMapper: ObjectMapper,
) {
    private val logger = LoggerFactory.getLogger(CardStatementScheduler::class.java)

    @Scheduled(cron = "0 0 4 1 * *") // 04:00 on the 1st of every month
    fun enqueueMonthlyCardStatement() {
        val yearMonth = YearMonth.now().toString()
        val dedupId = "$TASK_TYPE-$yearMonth"
        runCatching {
            taskQueue.enqueue(
                taskType = TASK_TYPE,
                payload = objectMapper.writeValueAsString(mapOf("yearMonth" to yearMonth)),
                groupId = TASK_TYPE,
                deduplicationId = dedupId,
            )
        }.onFailure {
            logger
                .atError()
                .addKeyValue("year_month", yearMonth)
                .setCause(it)
                .log("Failed to enqueue the monthly card statement Task")
        }
    }

    companion object {
        const val TASK_TYPE = "card.send-statement"
    }
}
